# Exercise 02 -  OLAP Cubes - Roll Up and Drill Down

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [3]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [4]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

The sql extension is already loaded. To reload it, use:
  %reload_ext sql
postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

## Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [6]:
%%time
%%sql
SELECT SUM(sales_amount) as revenue, rating, day, country
FROM factsales fa
JOIN dimmovie dm ON (dm.movie_key = fa.movie_key)
JOIN dimdate dd ON (dd.date_key = fa.date_key)
JOIN dimcustomer dc ON (dc.customer_key = fa.customer_key)
GROUP BY day, rating, country
ORDER BY revenue desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.46 ms, sys: 0 ns, total: 4.46 ms
Wall time: 29.3 ms


[(Decimal('169.67'), 'G', 30, 'China'),
 (Decimal('156.67'), 'PG', 30, 'India'),
 (Decimal('153.64'), 'NC-17', 30, 'India'),
 (Decimal('146.67'), 'PG-13', 30, 'China'),
 (Decimal('145.66'), 'R', 30, 'China'),
 (Decimal('143.68'), 'R', 30, 'India'),
 (Decimal('137.67'), 'G', 30, 'India'),
 (Decimal('135.75'), 'NC-17', 18, 'India'),
 (Decimal('131.72'), 'PG', 30, 'China'),
 (Decimal('128.74'), 'PG-13', 21, 'India'),
 (Decimal('121.72'), 'PG-13', 18, 'India'),
 (Decimal('119.76'), 'PG', 18, 'India'),
 (Decimal('116.72'), 'PG-13', 30, 'India'),
 (Decimal('115.77'), 'NC-17', 21, 'China'),
 (Decimal('115.75'), 'R', 21, 'India'),
 (Decimal('115.72'), 'NC-17', 27, 'India'),
 (Decimal('111.75'), 'PG-13', 17, 'China'),
 (Decimal('109.76'), 'NC-17', 17, 'United States'),
 (Decimal('108.77'), 'NC-17', 30, 'China'),
 (Decimal('101.79'), 'PG-13', 20, 'India')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

## Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

TODO: Write a query that calculates revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results. The first few rows of your output should match the table below.

In [7]:
%%time
%%sql
SELECT SUM(sales_amount) as revenue, rating, day, district
FROM factsales fa
JOIN dimmovie dm ON (dm.movie_key = fa.movie_key)
JOIN dimdate dd ON (dd.date_key = fa.date_key)
JOIN dimcustomer dc ON (dc.customer_key = fa.customer_key)
GROUP BY day, rating, district
ORDER BY revenue desc
LIMIT 20;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.21 ms, sys: 67 µs, total: 4.27 ms
Wall time: 31.6 ms


[(Decimal('53.88'), 'PG-13', 30, 'Southern Tagalog'),
 (Decimal('38.93'), 'G', 30, 'Inner Mongolia'),
 (Decimal('36.93'), 'G', 30, 'Shandong'),
 (Decimal('36.92'), 'NC-17', 30, 'West Bengali'),
 (Decimal('34.95'), 'PG-13', 17, 'Shandong'),
 (Decimal('32.94'), 'PG', 1, 'California'),
 (Decimal('32.93'), 'NC-17', 18, 'So Paulo'),
 (Decimal('31.93'), 'NC-17', 30, 'Buenos Aires'),
 (Decimal('31.93'), 'R', 21, 'So Paulo'),
 (Decimal('30.94'), 'PG', 30, 'Southern Tagalog'),
 (Decimal('30.93'), 'PG', 30, 'So Paulo'),
 (Decimal('30.92'), 'R', 30, 'Buenos Aires'),
 (Decimal('29.95'), 'G', 30, 'California'),
 (Decimal('29.95'), 'NC-17', 18, 'Maharashtra'),
 (Decimal('29.94'), 'PG-13', 21, 'Uttar Pradesh'),
 (Decimal('29.93'), 'PG', 20, 'Shandong'),
 (Decimal('29.93'), 'R', 10, 'Maharashtra'),
 (Decimal('28.97'), 'R', 18, 'Sumy'),
 (Decimal('28.96'), 'PG-13', 21, 'Southern Tagalog'),
 (Decimal('28.95'), 'NC-17', 29, 'So Paulo')]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>